service/continuum/continuum.go

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) 2025 Hemi Labs, Inc.
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package continuum
+
+import (
+	"context"
+	"crypto/ecdh"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/juju/loggo"
+	"github.com/mitchellh/go-homedir"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/hemilabs/heminetwork/v2/service/deucalion"
+	"github.com/hemilabs/heminetwork/v2/service/pprof"
+)
+
+const (
+	logLevel = "INFO"
+	appName  = "cnt"
+
+	defaultListenAddress  = "localhost:45067"
+	defaultMaxConnections = 8
+)
+
+var log = loggo.GetLogger(appName)
+
+func init() {
+	if err := loggo.ConfigureLoggers(logLevel); err != nil {
+		panic(err)
+	}
+}
+
+type Config struct {
+	Home                    string
+	LogLevel                string
+	PprofListenAddress      string
+	PrometheusListenAddress string
+	PrometheusNamespace     string
+	Secret                  string
+	ListenAddress           string
+	MaxConnections          int
+}
+type Server struct {
+	mtx sync.RWMutex
+	wg  sync.WaitGroup
+
+	cfg *Config
+
+	// Sessions
+	sessions map[string]*Transport
+
+	// Secrets
+	secret *Secret
+
+	// Listener
+	listenConfig *net.ListenConfig
+
+	// Prometheus
+	promCollectors  []prometheus.Collector
+	promPollVerbose bool // set to true to print stats during poll
+	isRunning       bool
+}
+
+type Info struct {
+	Online bool
+}
+
+func NewDefaultConfig() *Config {
+	return &Config{
+		LogLevel:            logLevel,
+		PrometheusNamespace: appName,
+		Secret:              "",
+		ListenAddress:       defaultListenAddress,
+		MaxConnections:      defaultMaxConnections,
+	}
+}
+
+func NewServer(cfg *Config) (*Server, error) {
+	if cfg == nil {
+		cfg = NewDefaultConfig()
+	}
+	return &Server{
+		cfg:          cfg,
+		listenConfig: &net.ListenConfig{},
+	}, nil
+}
+
+func genID() string {
+	buf := make([]byte, 16)
+	if _, err := rand.Read(buf); err != nil {
+		panic(fmt.Errorf("read random: %w", err))
+	}
+	return hex.EncodeToString(buf)
+}
+
+func (s *Server) newSession(t *Transport) string {
+	for {
+		id := genID()
+		s.mtx.Lock()
+		if _, ok := s.sessions[id]; ok {
+			// ID is already used, retry.
```

Comment:
bug: `s.mtx` is not unlocked before `continue`, resulting in a deadlock.
```suggestion
			// ID is already used, retry.
			s.mtx.Unlock()
```